viper.go

```diff
@@ -588,36 +589,58 @@ func (v *Viper) searchMap(source map[string]interface{}, path []string) interfac
 // in their keys).
 //
 // Note: This assumes that the path entries and map keys are lower cased.
-func (v *Viper) searchMapWithPathPrefixes(source map[string]interface{}, path []string) interface{} {
+func (v *Viper) searchMapWithPathPrefixes(source interface{}, path []string) interface{} {
 	if len(path) == 0 {
 		return source
 	}
 
 	// search for path prefixes, starting from the longest one
 	for i := len(path); i > 0; i-- {
 		prefixKey := strings.ToLower(strings.Join(path[0:i], v.keyDelim))
+		if sourceSlice, ok := source.([]interface{}); ok {
```

Comment:
How about making this a type switch instead?

I'd also consider moving the bodies of these conditionals to functions like `searchSliceWithPathPrefixes` and `searchMapWithPathPrefixes`, so that this function is a bit shorter and better separated.